README.pod
lib/Data/Petitcom.pm

Changes
Revision history for Perl extension Data::Petitcom

0.02
        - support Windows

0.01
        - initial version


t/util.pl
use 5.10.0;
use strict;
use warnings;

use File::Basename;
use Path::Class;

sub LoadData {
    my $filename = shift;
    my $file = file( dirname(__FILE__), '_data', $filename );
    return scalar( $file->slurp );
}

1;


lib/Data/Petitcom.pm
package Data::Petitcom;

use 5.10.0;
use strict;
use warnings;
our $VERSION = '0.03';

use parent qw{ Exporter };
our @EXPORT      = qw{ Save Load QRCode };
our @EXPORT_OK   = qw{ SaveFile LoadFile };
our %EXPORT_TAGS = ( all => [qw{ Save Load QRCode SaveFile LoadFile }] );

use Carp ();
use Data::Petitcom::PTC;
use Data::Petitcom::QRCode;
use Data::Petitcom::Resource;
use Path::Class;

sub Save {
    my $raw_data     = shift;
    my %opts         = @_;
    my $resource     = delete $opts{resource} || 'PRG';
    my $obj_resource = get_resource( resource => $resource, %opts );
    $obj_resource->data($raw_data);
    my $obj_ptc = $obj_resource->save(%opts);
    return $obj_ptc->dump;
}

sub Load {
    my $raw_ptc      = shift;
    my %opts         = @_;
    my $obj_ptc      = Data::Petitcom::PTC->new->load($raw_ptc);
    my $obj_resource = $obj_ptc->restore(%opts);
    return $obj_resource->data;
}

sub QRCode {
    my $ptc        = shift;
    my %opts       = @_;
    my $qr_type    = delete $opts{type};
    my $qr_ecc     = delete $opts{ecc};
    my $qr_version = delete $opts{version};

    my $raw_ptc = undef;
    given ($ptc) {
        when ( ref $ptc eq 'Data::Petitcom::PTC' ) {
            $raw_ptc = $ptc->dump;
        }
        when ( ! Data::Petitcom::PTC->is_ptc($ptc) ) {
            $raw_ptc = Save($ptc, %opts);
        }
        default { $raw_ptc = $ptc }
    }

    my $qr = Data::Petitcom::QRCode->new(
        type    => $qr_type,
        ecc     => $qr_ecc,
        version => $qr_version
    );
    return $qr->plot($raw_ptc);
}

sub SaveFile {
    my $save_file = shift;
    my $raw_data  = shift;
    my $save_fh
        = ( ref $save_file eq 'GLOB' )
        ? $save_file
        : file($save_file)->open('>') || Carp::croak "open failed: $!";
    my $raw_ptc = Save($raw_data, @_);
    binmode $save_fh;
    print $save_fh $raw_ptc;
}

sub LoadFile {
    my $load_file = shift;
    my $load_fh
        = ( ref $load_file eq 'GLOB' )
        ? $load_file
        : file($load_file)->open('<') || Carp::croak "open failed: $!";
    binmode $load_fh;
    my $raw_data = Load( do { local $/; <$load_fh> }, @_ );
    return $raw_data;
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Data::Petitcom - Data converter for プチコンmkII / Petit Computer

=head1 SYNOPSIS

  use Data::Petitcom;

  my $prg = <<'EOP';
  PRINT "Hello World"
  EOP
  my $qrcodes = QRCode($prg, type => 'term', version => 5);
  my $qr_num = @$qrcodes;
  for my $i ( 1 .. $qr_num ) {
      printf(
          "QR: %d / %d\n%s\n",
          $i,
          $qr_num,
          $qrcodes->[ $i - 1 ],
      );
  }


  use Path::Class;
  use GD::Tiler;
  my $bmp = '8b_256x192.bmp';
  my $raw_bmp = file($bmp)->slurp;
  my $qrcodes = QRCode($raw_bmp, resource => 'GRP', name => 'TEST_GRP', type => 'img');
  my $tiled = GD::Tiler->tile(
      Images       => $qrcodes,
      Background   => 'white',
      ImagesPerRow => 3,
  );
  file('qrcodes.png')->open('>')->print($tiled);

=head1 DESCRIPTION

Data::Petitcom is data converter for プチコンmkII / Petit Computer.

following resources supported.

=over

=item PRG: Program

=item GRP: Graphics (256x192 pixels 8bit-color bitmap)

=item CHR: User character (256x64 pixels 8bit-color bitmap)

=item COL: Color (8bit-color bitmap)

=back

=head1 FUNCTIONS

=head2 Exported Functions

=over 4

=item Save($raw_data, %opts)

Turn raw data into PTC.

=item Load($raw_ptc, %opts)

Turn PTC into raw data.

=item QRCode(($raw_ptc|$raw_data|$obj_ptc), %opts)

Returns barcode for the data (or object) specified.

=back

=head2 Exportable Functions

=over 4

=item SaveFile($ptc_file, $raw_data, %opts)

Writes the PTC to a file.

=item LoadFile($ptc_file, %opts)

Reads the PTC from a file.

=back

=head1 AUTHOR

hayajo <hayajo@cpan.org>

=head1 SEE ALSO

L<プチコンmkII|http://smileboom.com/special/ptcm2/>

L<Petit Computer|http://www.petitcomputer.com/>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Data/Petitcom/PTC.pm
package Data::Petitcom::PTC;

use strict;
use warnings;
use bytes ();

use Carp ();
use Digest::MD5;
use List::Util;
use Data::Petitcom::Resource;

use constant PTC                => 'PETITCOM';
use constant PTC_NAME_MAXLENGTH => 8;
use constant PTC_SIGNATURE      => 'PX01';
use constant PTC_VERSION        => [ 'PETC0100', 'PETC0300' ];
use constant PTC_RESOURCE => {
    PRG => 0x00,
    GRP => 0x02,
    CHR => 0x03,
    COL => 0x05,
};

use constant PTC_OFFSET_RESOURCE => 0x08;
use constant PTC_OFFSET_NAME     => 0x0C;
use constant PTC_OFFSET_VERSION  => 0x24;
use constant PTC_OFFSET_DATA     => 0x30;

my %defaults = (
    data     => '',
    resource => 'PRG',
    version  => 'PETC0300',
    name     => 'DPTC',
);

sub new {
    my $class = ref $_[0] ? ref shift : shift;
    my $self = bless {}, $class;
    $self->init(@_) if ( $self->can('init') );
    return $self;
}

sub init {
    my $self = shift;
    my %args = @_;
    for ( keys %defaults ) {
        my $value = $args{$_} || $defaults{$_};
        ( $self->can($_) ) ? $self->$_($value) : ( $self->{$_} = $value );
    }
    return $self;
}

sub resource {
    my $self = shift;
    if (@_) {
        my $resource = uc(shift);
        ( defined PTC_RESOURCE->{$resource} )
            ? $self->{resource} = $resource
            : Carp::croak "unsupported resource: $resource";
    }
    return $self->{resource};
}

sub name {
    my $self = shift;
    if (@_) {
        my $name = shift;
        $name =~ s/\x00//g;
        Carp::croak "name allows '_0-9A-Z': $name"
            unless ( $name =~ /^[_0-9a-zA-Z]+$/ );
        $self->{name}
            = bytes::substr uc($name), 0, PTC_NAME_MAXLENGTH;
    }
    return $self->{name};
}

sub version {
    my $self = shift;
    if (@_) {
        my $version = uc(shift);
        $self->{version}
            = List::Util::first { $_ eq $version } @{ PTC_VERSION() }
            or Carp::croak "unsupported version: $version";
    }
    return $self->{version};
}

sub data {
    my $self = shift;
    $self->{data} = shift if ( @_ > 0 );
    my $format = $self->version . uc( 'R' . $self->resource );
    return $format . $self->{data};
}

sub dump {
    my $self = shift;
    my $header .= PTC_SIGNATURE;
    $header .= pack 'I', bytes::length( $self->data );
    $header .= pack 'I', PTC_RESOURCE->{ $self->resource };
    $header .= bytes::substr $self->name . "\x00" x PTC_NAME_MAXLENGTH, 0, PTC_NAME_MAXLENGTH;
    $header .= Digest::MD5::md5(PTC . $self->data);
    my $raw_ptc = $header . $self->data;
    return $raw_ptc;
}

sub load {
    my $self    = ref $_[0] ? shift : shift->new;
    my $raw_ptc = shift;
    Carp::croak "unsupported data:" unless ( $self->is_ptc($raw_ptc) );
    my $r_int = unpack 'I', bytes::substr( $raw_ptc, PTC_OFFSET_RESOURCE, 4 );
    $self->resource(List::Util::first { PTC_RESOURCE->{$_} == $r_int } keys %{ PTC_RESOURCE() });
    $self->name( bytes::substr $raw_ptc, PTC_OFFSET_NAME, 8 );
    $self->version( bytes::substr $raw_ptc, PTC_OFFSET_VERSION, 8 );
    $self->data( bytes::substr $raw_ptc, PTC_OFFSET_DATA );
    return $self;
}

sub restore {
    my $self = shift;
    my $resource = get_resource( resource => $self->resource );
    $resource->load($self, @_);
    return $resource;
}

sub is_ptc {
    my $class = shift;
    return 1 if ( bytes::substr( $_[0], 0, 4 ) eq PTC_SIGNATURE );
}

1;


lib/Data/Petitcom/Resource/COL.pm
package Data::Petitcom::Resource::COL;

use strict;
use warnings;

use parent qw{ Data::Petitcom::Resource };

use bytes ();
use Data::Petitcom::PTC;
use Data::Petitcom::BMP qw{ Load Dump RGB555toRGB888 RGB888toRGB555 };

use constant RESOUECE            => 'COL';
use constant PTC_OFFSET_DATA     => 0x0C;
use constant PTC_COLORMAP_LENGTH => 512;     # RGB555(16bit) x 256colors
use constant COLORMAP_COLS       => 16;


sub save {
    my $self = shift;
    my %opts = @_;
    my $name = delete $opts{name};
    my $ptc  = Data::Petitcom::PTC->new(
        resource => RESOUECE,
        name     => $name,
        version  => 'PETC0100',
    );
    my $bmp    = Load( $self->data );
    my $rgb555 = RGB888toRGB555( $bmp->{colormap} );
    $ptc->data( pack 'v*', @$rgb555 );
    return $ptc;
}

sub load {
    my $self = shift;
    my $ptc  = shift;
    my $raw_colormap
        = bytes::substr( $ptc->data, PTC_OFFSET_DATA, PTC_COLORMAP_LENGTH );
    my @rgb555  = unpack 'v*', $raw_colormap;    # LE-short
    my $rgb888  = RGB555toRGB888(@rgb555);
    my $raw_bmp = Dump(
        width    => COLORMAP_COLS,
        height   => COLORMAP_COLS,
        colormap => $rgb888,
        pixels => [ map {
            ( (COLORMAP_COLS - 1) - int( $_ / COLORMAP_COLS ) ) * COLORMAP_COLS + $_ % COLORMAP_COLS
        } ( 0 .. 255 ) ],
    );
    $self->data($raw_bmp);
    return $self;
}

1;


lib/Data/Petitcom/Resource/GRP.pm
package Data::Petitcom::Resource::GRP;

use strict;
use warnings;

use parent qw{ Data::Petitcom::Resource };

use bytes ();
use Data::Petitcom::PTC;
use Data::Petitcom::BMP qw{ Load BMP2DATA DATA2BMP };

use constant RESOURCE   => 'GRP';
use constant BMP_WIDTH  => 256;
use constant BMP_HEIGHT => 192;

sub data {
    my $self  = shift;
    if (my $raw_bmp = shift) {
        my $bmp = Load($raw_bmp);
        Carp::croak "unsupported width x height: $bmp->{width} x $bmp->{height}"
            if (   $bmp->{width} != $self->BMP_WIDTH()
                || $bmp->{height} != $self->BMP_HEIGHT() );
        $self->{data} = $raw_bmp;
    }
    return $self->{data};
}

sub save {
    my $self = shift;

    my %opts     = @_;
    my $name     = delete $opts{name};
    my $sp_width = ( $self->RESOURCE() eq 'GRP' )
        ? 64
        : delete $opts{sp_width} || 16;
    my $sp_height = ( $self->RESOURCE() eq 'GRP' )
        ? 64
        : delete $opts{sp_height} || 16;


    my $ptc = Data::Petitcom::PTC->new(
        resource => $self->RESOURCE(),
        name     => $name,
        version  => 'PETC0100',
        data     => BMP2DATA(
            $self->data,
            sp_width  => $sp_width,
            sp_height => $sp_height,
        ),
    );
    return $ptc;
}

sub load {
    my $self  = shift;
    my $ptc   = shift;

    my %opts     = @_;
    my $sp_width = ( $self->RESOURCE() eq 'GRP' )
        ? 64
        : delete $opts{sp_width} || 16;
    my $sp_height = ( $self->RESOURCE() eq 'GRP' )
        ? 64
        : delete $opts{sp_height} || 16;

    my $bmp  = DATA2BMP(
        bytes::substr( $ptc->data, 0x0C ),
        width     => $self->BMP_WIDTH(),
        height    => $self->BMP_HEIGHT(),
        sp_width  => $sp_width,
        sp_height => $sp_height,
    );
    $self->data($bmp);
    return $self;
}

1;


lib/Data/Petitcom/Resource/PRG.pm
package Data::Petitcom::Resource::PRG;

use strict;
use warnings;

use parent qw{ Data::Petitcom::Resource };

use Encode ();
use Unicode::Japanese;
use Data::Petitcom::CharTBL qw{ dump_char load_char };
use Data::Petitcom::PTC;
use bytes ();

use constant RESOUECE        => 'PRG';
use constant PTC_OFFSET_CODE => 0x18;

sub save {
    my $self = shift;
    my %opts = @_;
    my $name     = delete $opts{name};
    my $encoding = delete $opts{encoding};
    my $ptc = $self->_create_ptc($name, $encoding);
    return $ptc;
}

sub load {
    my $self = shift;
    my ( $ptc, %opts ) = @_;
    my $zenkaku  = delete $opts{zenkaku};
    my $encoding = delete $opts{encoding};
    my $code = substr $ptc->data, PTC_OFFSET_CODE;
    $self->data( _decode( $code, $zenkaku, $encoding ) );
    return $self;
}

sub _create_ptc {
    my $self = shift;
    my ($name, $encoding) = @_;
    my $ptc  = Data::Petitcom::PTC->new(
        resource => RESOUECE,
        name     => $name,
    );
    my $code = _encode( $self->data, $encoding );
    my $header_data = pack 'C*', 0x00, 0x00, 0x00, 0x00;
    $header_data .= pack 'C*', 0x00, 0x00, 0x00, 0x00;
    $header_data .= pack 'I', bytes::length( $code );
    $ptc->data( $header_data . $code );
    return $ptc;
}

sub _encode {
    my $code     = shift;
    my $encoding = shift;
    $code = ($encoding)
        ? Encode::find_encoding($encoding)->decode($code)
        : Encode::decode_utf8($code);
    $code = Unicode::Japanese->new($code)->hira2kata->z2h->getu;
    $code =~ s/\r\n/\r/g;
    $code =~ s/\n/\r/g;
    my $encoded = '';
    for my $i ( 0 .. ( length($code) - 1 ) ) {
        my $char = substr( $code, $i, 1 );
        $encoded .= dump_char($char);
    }
    return $encoded;
}

sub _decode {
    my $binary = shift;
    my ( $zenkaku, $encoding ) = @_;
    my $decoded = '';
    for my $i ( 0 .. ( bytes::length($binary) - 1 ) ) {
        my $byte = bytes::substr( $binary, $i, 1 );
        $decoded .= load_char($byte);
    }
    if ($zenkaku) {
        # $decoded = Unicode::Japanese->new($decoded)->h2zKanaK->get;
        # $decoded = Unicode::Japanese->new($decoded)->h2z->get;
        $decoded = Unicode::Japanese->new($decoded)->h2zKanaK->h2z->getu;
    }
    if ($encoding) {
        Encode::from_to( $decoded, 'utf8', $encoding );
    }
    return $decoded;
}

1;


lib/Data/Petitcom/QRCode.pm
use strict;
use warnings;

{
    package Data::Petitcom::QRCode;

    use 5.10.0;
    use bytes();

    use base qw{ Exporter };
    our @EXPORT_OK = qw{ plot_qrcode };

    use Carp ();
    use Compress::Zlib ();
    use Digest::MD5;
    use POSIX qw{ ceil floor };
    use GD::Barcode::QRcode;

    use constant PTC_OFFSET_FILENAME     => 0x0C;
    use constant PTC_OFFSET_DATA         => 0x24;
    use constant PTC_OFFSET_RESOURCENAME => 0x2C;

    use constant PTC_QR_SIGNATURE                => 'PT';
    use constant DEFAULT_PTC_QR_VERSION_IMAGE    => 20;
    use constant DEFAULT_PTC_QR_VERSION_TERM     => 4;
    use constant DEFAULT_PTC_QR_IMAGE_MODULESIZE => 5;

    use constant QR_ECC => +{ L => 0, M => 1, Q => 2, H => 3 };
    use constant QR_VERSION => [
        [ 0,    0,    0,    0 ],
        [ 17,   14,   11,   7 ],
        [ 32,   26,   20,   14 ],
        [ 53,   42,   32,   24 ],
        [ 78,   62,   46,   34 ],
        [ 106,  84,   60,   44 ],
        [ 134,  106,  74,   58 ],
        [ 154,  122,  86,   64 ],
        [ 192,  152,  108,  84 ],
        [ 230,  180,  130,  98 ],
        [ 271,  213,  151,  119 ],
        [ 321,  251,  177,  137 ],
        [ 367,  287,  203,  155 ],
        [ 425,  331,  241,  177 ],
        [ 458,  362,  258,  194 ],
        [ 520,  412,  292,  220 ],
        [ 586,  450,  322,  250 ],
        [ 644,  504,  364,  280 ],
        [ 718,  560,  394,  310 ],
        [ 792,  624,  442,  338 ],
        [ 858,  666,  482,  382 ],
        [ 929,  711,  509,  403 ],
        [ 1003, 779,  565,  439 ],
        [ 1091, 857,  611,  461 ],
        [ 1171, 911,  661,  511 ],
        [ 1273, 997,  715,  535 ],
        [ 1367, 1059, 751,  593 ],
        [ 1465, 1125, 805,  625 ],
        [ 1528, 1190, 868,  658 ],
        [ 1628, 1264, 908,  698 ],
        [ 1732, 1370, 982,  742 ],
        [ 1840, 1452, 1030, 790 ],
        [ 1952, 1538, 1112, 842 ],
        [ 2068, 1628, 1168, 898 ],
        [ 2188, 1722, 1228, 958 ],
        [ 2303, 1809, 1283, 983 ],
        [ 2431, 1911, 1351, 1051 ],
        [ 2563, 1989, 1423, 1093 ],
        [ 2699, 2099, 1499, 1139 ],
        [ 2809, 2213, 1579, 1219 ],
        [ 2953, 2331, 1663, 1273 ],
    ];

    my %defaults = (
        type    => 'text',
        ecc     => 'M',
        version => DEFAULT_PTC_QR_VERSION_IMAGE
    );
    eval "sub $_ { \@_ > 1 ? \$_[0]->{$_} = \$_[1] : \$_[0]->{$_} }" for keys %defaults;

    sub new {
        my $class = ref $_[0] ? ref shift : shift;
        my $self = bless {@_}, $class;
        $self->init() if ( $self->can('init') );
        return $self;
    }

    sub init {
        my $self = shift;
        for ( keys %defaults ) {
            my $value = $self->{$_} || $defaults{$_};
            ( $self->can($_) ) ? $self->$_($value) : ( $self->{$_} = $value );
        }
        return $self;
    }

    sub plot_qrcode {
        my ($ptc, %opts) = @_;
        my $qrcode = __PACKAGE__->new(%opts);
        return $qrcode->plot($ptc);
    }

    sub plot {
        my $self    = shift;
        my $raw_ptc = shift;
        return $self->_generate_qrcode($raw_ptc);
    }

    sub _generate_qrcode {
        my $self    = shift;
        my $raw_ptc = shift;

        my $plot_type = $self->type;
        my $ecc       = $self->ecc;
        my $version   = $self->version;

        my $qr_bin           = _create_qr_bin($raw_ptc);
        my $max_qr_data_size = _max_qr_data_size( $version, QR_ECC->{$ecc} );
        my $number_of_qr     = _number_of_qr( bytes::length($qr_bin), $max_qr_data_size );
        my $qr_opts          = {
            Ecc        => $ecc,
            Version    => $version,
            ModuleSize => ( $plot_type eq 'image' )
            ? DEFAULT_PTC_QR_IMAGE_MODULESIZE
            : 1
        };

        my @qrcode = ();
        for my $count_qr ( 1 .. $number_of_qr ) {
            my $a_qr_data = bytes::substr( $qr_bin, ( $count_qr - 1 ) * $max_qr_data_size, $max_qr_data_size );

            my $a_qr_bin = PTC_QR_SIGNATURE;
            $a_qr_bin .= pack 'C', $count_qr;
            $a_qr_bin .= pack 'C', $number_of_qr;
            $a_qr_bin .= Digest::MD5::md5($a_qr_data);
            $a_qr_bin .= Digest::MD5::md5($qr_bin);
            $a_qr_bin .= $a_qr_data;

            my $qrcode = do {
                given ($plot_type) {
                    when ('term') {
                        GD::Barcode::QRcode::Text->new( $a_qr_bin, $qr_opts )->term;
                    }
                    when ('image') {
                        my $gd = GD::Barcode::QRcode->new( $a_qr_bin, $qr_opts )->plot;
                        $gd->string(
                            GD::Font->Large,
                            5, 2,    # 0, 0 => left-top
                            "$count_qr / $number_of_qr",
                            $gd->colorAllocate( 0, 0, 0 ),    # black
                        );
                        $gd->png;
                    }
                    default {
                        GD::Barcode::QRcode::Text->new( $a_qr_bin, $qr_opts )->barcode;
                    }
                }
            };
            push @qrcode, $qrcode;
        }

        return \@qrcode;
    }

    sub _deflate_data {
        my $code     = shift;
        my $deflater = Compress::Zlib::deflateInit()
            or Carp::croak "deflateInit() failed: $!";

        my $zdata = $deflater->deflate($code);
        $zdata .= $deflater->flush();

        return $zdata;
    }

    sub _create_qr_bin {
        my $raw_ptc = shift || return;

        my $filename = unpack 'Z*', bytes::substr( $raw_ptc, PTC_OFFSET_FILENAME, 8 );
        my $resource = bytes::substr( $raw_ptc, PTC_OFFSET_RESOURCENAME, 4 );
        my $data     = bytes::substr( $raw_ptc, PTC_OFFSET_DATA );
        my $zdata    = _deflate_data($data);

        my $qr_bin = bytes::substr( $filename . "\x00" x 8, 0, 8 );
        $qr_bin .= $resource;
        $qr_bin .= pack 'I', bytes::length($zdata);
        $qr_bin .= pack 'I', ( bytes::length($raw_ptc) - PTC_OFFSET_DATA );
        $qr_bin .= $zdata;

        return $qr_bin;
    }

    sub _max_qr_data_size {
        my $version = shift || DEFAULT_PTC_QR_VERSION_TERM;
        my $ecc     = shift // 1;

        Carp::croak "version between 1 and 24"
            if ( $version < 1 || $version > 24 );
        Carp::croak "ecc between 0 and 3"
            if ( $ecc < 0 || $ecc > 3 );
        Carp::croak "invalid combination of version x ecc: $version x $ecc"
            if (QR_VERSION->[$version]->[$ecc] <= PTC_OFFSET_DATA);

        return QR_VERSION->[$version]->[$ecc] - PTC_OFFSET_DATA;
    }

    sub _number_of_qr {
        my ( $total_size, $part_size ) = @_;
        return floor( $total_size / $part_size ) + 1;
    }
}

{
    package GD::Barcode::QRcode::Text;

    use parent qw{ GD::Barcode::QRcode };
    use bytes ();
    use Term::ANSIColor;

    sub new {
        my $class = shift;
        my $self  = $class->SUPER::new(@_);
        bless $self, $class;
        return $self;
    }

    sub barcode {
        my $self = shift;
        return _trim_margin( $self->SUPER::barcode() );
    }

    sub term {
        my $self      = shift;
        my $qr_text   = $self->barcode;
        my $term_text = '';

        for my $i ( 0 .. ( bytes::length($qr_text) - 1 ) ) {
            my $module = bytes::substr $qr_text, $i, 1;
            $term_text .=
                ( $module =~ /^[01]$/ )
                ? colored( '  ', ($module) ? 'on_black' : 'on_white' ) # 2 spaces width
                : $module;
        }

        return $term_text;
    }

    sub _trim_margin {
        my $qr_text   = shift;
        my $qr_aryref = [ map { [ split //, $_ ] } split /\n/, $qr_text ];
        my $ret_text  = '';

        for my $i ( 3 .. ( ( @$qr_aryref - 1 ) - 3 ) ) {
            my $line = $qr_aryref->[$i];
            for my $j ( 3 .. ( ( @$line - 1 ) - 3 ) ) {
                $ret_text .= $line->[$j];
            }
            $ret_text .= "\n";
        }

        return $ret_text;
    }
}

1;
